gui-agent/vmside.c

```diff
@@ -2279,6 +2379,66 @@ int main(int argc, char **argv)
     int xfd;
     Ghandles g;
     int wait_fds[2];
+    
+    g.created_input_device = access("/run/qubes-service/gui-agent-virtual-input-device", F_OK) == 0;
+
+    if(g.created_input_device) {
+        // open uinput, if it fails, falls back to xdriver to not break input to the qube
+        g.uinput_fd = open("/dev/uinput", O_WRONLY | O_NDELAY);
+        if(g.uinput_fd < 0) {
+            g.uinput_fd = open("/dev/input/uinput", O_WRONLY | O_NDELAY);
+            
+            if(g.uinput_fd < 0) {
+                fprintf(stderr, "Couldn't open uinput, falling back to xdriver\n");
+                g.created_input_device = 0;
+            }
+        }
+    }
+    
+    // input device creation
+    if(g.created_input_device) {
+        
+        if (ioctl(g.uinput_fd, UI_SET_EVBIT, EV_SYN) < 0) {
+            fprintf(stderr, "error setting EVBIT for EV_SYN, falling back to xdriver\n");
+            g.created_input_device = 0;
+        }
+
+        if (ioctl(g.uinput_fd, UI_SET_EVBIT, EV_KEY) < 0) {
+            fprintf(stderr, "error setting EVBIT for EV_KEY, falling back to xdriver\n");
+            g.created_input_device = 0;
+        }
+        
+
+        
+        
+
+        // set all keys
+        for(int i = 1; i < KEY_MAX; i++) {
+            if(i != KEY_RESERVED && ioctl(g.uinput_fd, UI_SET_KEYBIT, i) < 0) {
```

Comment:
With all keys set, udev (and so Xorg) detects this as a combo keyboard+mouse device now. See `udevadm info -q all -n /dev/input/event1` (or whatever device node it will get). It should not have `ID_INPUT_MOUSE` set. I guess you need to exclude `BTN_*` range (`BTN_MISC` .. `BTN_GEAR_UP` - various buttons of mouses, tablets, game pads etc)